### ISSUE TITLE
completing-read-multiple with helm-mode won't let me mark several items

### ISSUE THREAD

**Issue by DamienCassou** [emacs-helm/helm#1692]
## Expected behavior

When using helm-mode and calling `completing-read-multiple`, I should be able to mark several items. Beyound manually typing a comma to separate items, I didn't find a way to get `completing-read-multiple` to return several items with `helm-mode`.

## Steps to reproduce (recipe)

- `./emacs-helm.sh`
- `(completing-read-multiple "Select items: " '(a b c d))`

I would like to mark items here.

## Describe versions of helm, emacs, operating system etc.

- helm: 2.5.2
- emacs: 25.1.1

**Comment by thierryvolpiatto** [comment]
Damien Cassou <notifications@github.com> writes:

> Expected behavior
>
> When using helm-mode and calling completing-read-multiple, I should be
> able to mark several items. Beyound manually typing a comma to separate
> items, I didn't find a > way to get completing-read-multiple to return
> several items with helm-mode.

You can't do that like BTW completing-read, because these are helmized
functions that behave like emacs vanilla, i.e if your completing-read
returns '("a") instead of "a" your native emacs function will not be
happy as it have been builded to return "a".
So no I wont implement this, the current behavior is correct.

--
Thierry

**Comment by thierryvolpiatto** [comment]
Damien Cassou <notifications@github.com> writes:

>   * ./emacs-helm.sh
>   * (completing-read-multiple "Select items: " '(a b c d))
>
> I would like to mark items here.

I think what you really want is something like this:

    (mapconcat 'identity (helm-comp-read "test: " '(a b c d) :marked-candidates t) ",")

-- 
Thierry

**Comment by xiongtx** [comment]
Instead of `(mapconcat 'identity ... ",")`, there's `string-join`.

**Comment by DamienCassou** [comment]
I don't understand what you are saying. Let me rephrase what I have in mind. In plain Emacs (`emacs -Q`),

    (completing-read-multiple "Select items: " '(a b c d))

returns a list (e.g., `("a" "d")`) if the user selects multiple items (through typing a `, TAB` between each choice).

I would like `helm-mode` to let me select several items without typing `, TAB` but through `helm-toggle-visible-mark` instead.

**Comment by thierryvolpiatto** [comment]
Damien Cassou <notifications@github.com> writes:

> I don't understand what you are saying.

I perfectly understood what you want and said it is not possible to
achieve this See
https://github.com/emacs-helm/helm/wiki/FAQ#can-you-modify-the-emacs-vanilla-helmized-command-to-behave-differently
I have also added an entry in helm help.

-- 
Thierry

**Comment by DamienCassou** [comment]
I understand that helm-mode should preserve both API and return value of Emacs vanilla. That's fine and I'm not asking to change that. `completing-read-multiple` returns a list of values in Emacs vanilla and I want to keep it that way in helm-mode. The thing I would like to see change in helm-mode is how the UI let the user choose the different element of the returned list.

Why are you proposing me to concatenate elements of the list into a string?

**Comment by thierryvolpiatto** [comment]
Damien Cassou <notifications@github.com> writes:

> I understand that helm-mode should preserve both API and return value
> of Emacs vanilla. That's fine and I'm not asking to change that.
> completing-read-multiple returns a list of values in Emacs vanilla and
> I want to keep it that way in helm-mode. The thing I would like to see
> change in helm-mode is how the UI let the user choose the different
> element of the returned list.

Would need a complete rewrite of the completion-in-region mechanism, a
big work for a small benefit IMO.

> Why are you proposing me to concatenate elements of the list into a string?

Forget it, I just wanted you use helm-comp-read with :marked-candidates
which have the same effect.

--
Thierry

**Comment by thierryvolpiatto** [comment]
Thierry Volpiatto <thierry.volpiatto@gmail.com> writes:

> Would need a complete rewrite of the completion-in-region mechanism, a
> big work for a small benefit IMO.

Otherwise an other alternative that can be easily implemented would be
to allow marking and insert the resulting list mapconcated with "," as
separator in the inner minibuffer.

-- 
Thierry
